File: src/main/java/me/dkim19375/dkim19375core/external/VaultJavaPlugin.java
package me.dkim19375.dkim19375core.external;

import me.dkim19375.dkim19375core.CoreJavaPlugin;
import net.milkbowl.vault.chat.Chat;
import net.milkbowl.vault.economy.Economy;
import net.milkbowl.vault.permission.Permission;
import org.bukkit.plugin.RegisteredServiceProvider;

public abstract class VaultJavaPlugin extends CoreJavaPlugin {
    protected static Economy econ = null;
    protected static Permission perms = null;
    protected static Chat chat = null;

    @Override
    public void onEnable() {
        if (!setupEconomy() ) {
            printToConsole("Disabled due to no Vault dependency found!");
            getServer().getPluginManager().disablePlugin(this);
            return;
        }
        setupPermissions();
        setupChat();
    }

    @Override
    public abstract void onDisable();

    protected boolean setupEconomy() {
        if (getServer().getPluginManager().getPlugin("Vault") == null) {
            return false;
        }
        RegisteredServiceProvider<Economy> rsp = getServer().getServicesManager().getRegistration(Economy.class);
        if (rsp == null) {
            return false;
        }
        econ = rsp.getProvider();
        return econ != null;
    }

    protected boolean setupChat() {
        RegisteredServiceProvider<Chat> rsp = getServer().getServicesManager().getRegistration(Chat.class);
        chat = rsp.getProvider();
        return chat != null;
    }

    protected boolean setupPermissions() {
        RegisteredServiceProvider<Permission> rsp = getServer().getServicesManager().getRegistration(Permission.class);
        perms = rsp.getProvider();
        return perms != null;
    }

    public static Economy getEconomy() {
        return econ;
    }

    public static Permission getPermissions() {
        return perms;
    }

    public static Chat getChat() {
        return chat;
    }
}

File: src/main/kotlin/me/dkim19375/dkim19375core/javaplugin/VaultJavaPlugin.kt
package me.dkim19375.dkim19375core.javaplugin

import me.dkim19375.dkim19375core.annotation.API
import net.milkbowl.vault.chat.Chat
import net.milkbowl.vault.economy.Economy
import net.milkbowl.vault.permission.Permission

@API
abstract class VaultJavaPlugin : CoreJavaPlugin() {
    @API
    lateinit var econ: Economy
    @API
    lateinit var permissions: Permission
    @API
    lateinit var chat: Chat

    override fun onEnable() {
        setupEconomy()
        setupPermissions()
        setupChat()
    }

    abstract override fun onDisable()

    @API
    protected fun setupEconomy(): Boolean {
        if (server.pluginManager.getPlugin("Vault") == null) {
            return false
        }
        val rsp = server.servicesManager.getRegistration(
            Economy::class.java
        ) ?: return false
        econ = rsp.provider
        return this::econ.isInitialized
    }

    @API
    protected fun setupChat(): Boolean {
        if (server.pluginManager.getPlugin("Vault") == null) {
            return false
        }
        val rsp = server.servicesManager.getRegistration(
            Chat::class.java
        ) ?: return false
        chat = rsp.provider
        return this::chat.isInitialized
    }

    @API
    protected fun setupPermissions(): Boolean {
        if (server.pluginManager.getPlugin("Vault") == null) {
            return false
        }
        val rsp = server.servicesManager.getRegistration(
            Permission::class.java
        ) ?: return false
        permissions = rsp.provider
        return this::permissions.isInitialized
    }
}